src/evo/mnhftx.cpp

```diff
@@ -80,6 +83,183 @@ bool CheckMNHFTx(const CTransaction& tx, const CBlockIndex* pindexPrev, TxValida
     return true;
 }
 
+static bool extractSignals(const CBlock& block, const CBlockIndex* const pindex, std::vector<uint8_t>& signals_to_process, BlockValidationState& state)
+{
+    AssertLockHeld(cs_main);
+
+    // we skip the coinbase
+    for (size_t i = 1; i < block.vtx.size(); ++i) {
+        const CTransaction& tx = *block.vtx[i];
+
+        if (tx.nVersion != 3 || tx.nType != TRANSACTION_MNHF_SIGNAL) {
+            // only interested in special TXs 'TRANSACTION_MNHF_SIGNAL'
+            continue;
+        }
+
+        TxValidationState tx_state;
+        if (!CheckMNHFTx(tx, pindex, tx_state)) {
+            return state.Invalid(BlockValidationResult::BLOCK_CONSENSUS, tx_state.GetRejectReason(), tx_state.GetDebugMessage());
+        }
+
+        MNHFTxPayload mnhfTx;
+        if (!GetTxPayload(tx, mnhfTx)) {
+            return state.Invalid(BlockValidationResult::BLOCK_CONSENSUS, "bad-mnhf-tx-payload");
+        }
+        signals_to_process.push_back(mnhfTx.signal.versionBit);
+    }
+
+    // Checking that there's no any duplicates...
+    std::sort(signals_to_process.begin(), signals_to_process.end());
+    const auto it = std::unique(signals_to_process.begin(), signals_to_process.end());
+    if (std::distance(signals_to_process.begin(), it) != signals_to_process.size()) {
+        return state.Invalid(BlockValidationResult::BLOCK_CONSENSUS, "bad-mnhf-duplicates");
+    }
+
+    return true;
+}
+
+bool CMNHFManager::ProcessBlock(const CBlock& block, const CBlockIndex* const pindex, bool fJustCheck, BlockValidationState& state)
+{
+    try {
+        std::vector<uint8_t> new_signals;
+        if (!extractSignals(block, pindex, new_signals, state)) {
+            // state is set inside extractSignals
+            return false;
+        }
+        if (new_signals.empty()) {
+            if (!fJustCheck) {
+                AddToCache(GetFromCache(pindex->pprev), pindex);
+            }
+            return true;
+        }
+
+        Signals signals = GetFromCache(pindex->pprev);
+        int mined_height = pindex->nHeight;
+
+        // Extra validation of signals to be sure that it can succeed
+        for (const auto& versionBit : new_signals) {
+            LogPrintf("%s: add mnhf bit=%d block:%s number of known signals:%lld\n", __func__, versionBit, pindex->GetBlockHash().ToString(), signals.size());
+            if (signals.find(versionBit) != signals.end()) {
+                return state.Invalid(BlockValidationResult::BLOCK_CONSENSUS, "bad-mnhf-duplicate");
+            }
+
+            if (!Params().UpdateMNActivationParam(versionBit, mined_height, pindex->GetMedianTimePast(), true /* fJustCheck */)) {
+                return state.Invalid(BlockValidationResult::BLOCK_CONSENSUS, "bad-mnhf-non-mn-fork");
```

Comment:
Won't old nodes simply fork off the moment a mnhf is mined because they have no idea about new deployment(s)?

Reply:
Hardfork "MN RR" is planned to happen after hard-fork "v20". It means that all nodes are ready for MnEHF tx

Reply:
yeah, but what happens when we try to activate another deployment with mnhf in the future?

Reply:
@UdjinM6 please check https://github.com/dashpay/dash/pull/5469/commits/5b2d6910fa3471b591c0abbc1a2e3ced92990b46 
Also functional tests [updated accordingly](https://github.com/dashpay/dash/pull/5469/commits/79849bc6041371a49c94e6382715d7f093525a5b)  and fixes for [validation's code](https://github.com/dashpay/dash/pull/5469/commits/644413e6aadc142c833d390dc4604abe183e1637)

I think this approach would work